File: Practica 1/r/02_analisis_datos.R
suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(ggplot2)
  library(tidyr)
})

args <- commandArgs(trailingOnly = TRUE)

ruta_csv <- ifelse(
  length(args) >= 1,
  args[1],
  "ventas_online_2025_limpio.csv"
)

salida <- ifelse(
  length(args) >= 2,
  args[2],
  "salida_r"
)

dir.create(salida, showWarnings = FALSE, recursive = TRUE)
dir.create(
  file.path(salida, "graficos"),
  showWarnings = FALSE,
  recursive = TRUE
)
dir.create(
  file.path(salida, "tablas"),
  showWarnings = FALSE,
  recursive = TRUE
)

# IMPORTANTE:
# Este script NO hace limpieza.
# Lee el archivo generado por 01_limpieza_datos.R.
df <- read_csv(ruta_csv, show_col_types = FALSE)

df$FechaCompra <- as.Date(df$FechaCompra)

df <- df |>
  mutate(
    Mes = format(FechaCompra, "%Y-%m"),
    Genero_desc = ifelse(
      Genero == 1,
      "Femenino",
      "Masculino"
    ),
    MetodoPago_desc = case_when(
      MetodoPago == 0 ~ "Efectivo",
      MetodoPago == 1 ~ "Tarjeta de Crédito",
      MetodoPago == 2 ~ "Tarjeta de Débito"
    ),
    Navegador_desc = case_when(
      Navegador == 0 ~ "Tienda Física",
      Navegador == 1 ~ "Navegador 1",
      Navegador == 2 ~ "Navegador 2",
      Navegador == 3 ~ "Navegador 3",
      Navegador == 4 ~ "Navegador 4"
    ),
    Boletin_desc = ifelse(Boletin == 1, "Sí", "No"),
    Vale_desc = ifelse(Vale == 1, "Sí", "No")
  )

moda <- function(x) {
  valores <- unique(x)
  valores[which.max(tabulate(match(x, valores)))]
}

# Estadísticas
columnas <- c(
  "Edad", "Venta_total", "N_Compras",
  "MontoCompra", "Tiempo"
)

estadisticas <- bind_rows(
  lapply(columnas, function(columna) {
    x <- df[[columna]]

    data.frame(
      Variable = columna,
      Media = mean(x),
      Mediana = median(x),
      Moda = moda(x),
      Desviacion_estandar = sd(x),
      Minimo = min(x),
      Maximo = max(x)
    )
  })
)

write_csv(
  estadisticas,
  file.path(salida, "tablas", "estadisticas_basicas.csv")
)

# Ventas por mes
ventas_mes <- df |>
  group_by(Mes) |>
  summarise(
    Ventas = sum(MontoCompra),
    .groups = "drop"
  ) |>
  arrange(Mes)

write_csv(
  ventas_mes,
  file.path(salida, "tablas", "ventas_por_mes.csv")
)

p <- ggplot(
  ventas_mes,
  aes(Mes, Ventas, group = 1)
) +
  geom_line() +
  geom_point() +
  labs(
    title = "Ventas por mes",
    x = "Mes",
    y = "Monto de ventas"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(
      angle = 45,
      hjust = 1
    )
  )

ggsave(
  file.path(salida, "graficos", "01_ventas_por_mes.png"),
  p,
  width = 10,
  height = 5,
  dpi = 160
)

# Método de pago
pago <- df |>
  group_by(MetodoPago_desc) |>
  summarise(
    Ventas = sum(MontoCompra),
    .groups = "drop"
  )

write_csv(
  pago,
  file.path(salida, "tablas", "ventas_por_metodo_pago.csv")
)

p <- ggplot(
  pago,
  aes(reorder(MetodoPago_desc, Ventas), Ventas)
) +
  geom_col() +
  coord_flip() +
  labs(
    title = "Ventas por método de pago",
    x = "Método de pago",
    y = "Monto de ventas"
  ) +
  theme_minimal()

ggsave(
  file.path(salida, "graficos", "02_ventas_por_metodo_pago.png"),
  p,
  width = 8,
  height = 5,
  dpi = 160
)

# Navegador
navegador <- df |>
  group_by(Navegador_desc) |>
  summarise(
    Ventas = sum(MontoCompra),
    .groups = "drop"
  )

write_csv(
  navegador,
  file.path(salida, "tablas", "ventas_por_navegador.csv")
)

p <- ggplot(
  navegador,
  aes(reorder(Navegador_desc, Ventas), Ventas)
) +
  geom_col() +
  coord_flip() +
  labs(
    title = "Ventas por navegador/canal",
    x = "Navegador / canal",
    y = "Monto de ventas"
  ) +
  theme_minimal()

ggsave(
  file.path(salida, "graficos", "03_ventas_por_navegador.png"),
  p,
  width = 9,
  height = 5,
  dpi = 160
)

# Boletín
boletin <- df |>
  group_by(Boletin_desc) |>
  summarise(
    Ventas = sum(MontoCompra),
    .groups = "drop"
  )

write_csv(
  boletin,
  file.path(salida, "tablas", "ventas_por_boletin.csv")
)

p <- ggplot(
  boletin,
  aes(Boletin_desc, Ventas)
) +
  geom_col() +
  labs(
    title = "Ventas según uso de boletín",
    x = "Boletín",
    y = "Monto de ventas"
  ) +
  theme_minimal()

ggsave(
  file.path(salida, "graficos", "04_ventas_por_boletin.png"),
  p,
  width = 7,
  height = 5,
  dpi = 160
)

# Vale
vale <- df |>
  group_by(Vale_desc) |>
  summarise(
    Ventas = sum(MontoCompra),
    .groups = "drop"
  )

write_csv(
  vale,
  file.path(salida, "tablas", "ventas_por_vale.csv")
)

p <- ggplot(
  vale,
  aes(Vale_desc, Ventas)
) +
  geom_col() +
  labs(
    title = "Ventas según uso de vale",
    x = "Vale",
    y = "Monto de ventas"
  ) +
  theme_minimal()

ggsave(
  file.path(salida, "graficos", "05_ventas_por_vale.png"),
  p,
  width = 7,
  height = 5,
  dpi = 160
)

# Venta total vs edad
p <- ggplot(
  df,
  aes(Edad, Venta_total)
) +
  geom_point(alpha = 0.6) +
  labs(
    title = "Venta total vs edad",
    x = "Edad",
    y = "Venta total"
  ) +
  theme_minimal()

ggsave(
  file.path(salida, "graficos", "06_venta_total_vs_edad.png"),
  p,
  width = 8,
  height = 5,
  dpi = 160
)

# Segmentación por edad
df <- df |>
  mutate(
    GrupoEdad = cut(
      Edad,
      breaks = c(
        -Inf, 17, 25, 35,
        45, 55, 65, Inf
      ),
      labels = c(
        "<=17", "18-25", "26-35",
        "36-45", "46-55", "56-65", "66+"
      )
    )
  )

edad <- df |>
  group_by(GrupoEdad) |>
  summarise(
    Compras = n(),
    Ventas = sum(MontoCompra),
    TicketPromedio = mean(MontoCompra),
    .groups = "drop"
  )

write_csv(
  edad,
  file.path(salida, "tablas", "segmentacion_por_edad.csv")
)

p <- ggplot(
  edad,
  aes(GrupoEdad, Ventas)
) +
  geom_col() +
  labs(
    title = "Ventas por grupo de edad",
    x = "Grupo de edad",
    y = "Monto de ventas"
  ) +
  theme_minimal()

ggsave(
  file.path(salida, "graficos", "07_ventas_por_grupo_edad.png"),
  p,
  width = 9,
  height = 5,
  dpi = 160
)

# Género
genero <- df |>
  group_by(Genero_desc) |>
  summarise(
    Compras = n(),
    Ventas = sum(MontoCompra),
    TicketPromedio = mean(MontoCompra),
    .groups = "drop"
  )

write_csv(
  genero,
  file.path(salida, "tablas", "comparacion_genero.csv")
)

p <- ggplot(
  genero,
  aes(Genero_desc, Ventas)
) +
  geom_col() +
  labs(
    title = "Ventas por género",
    x = "Género",
    y = "Monto de ventas"
  ) +
  theme_minimal()

ggsave(
  file.path(salida, "graficos", "08_ventas_por_genero.png"),
  p,
  width = 7,
  height = 5,
  dpi = 160
)

# Boletines y vales por mes
uso_mes <- df |>
  group_by(Mes) |>
  summarise(
    Boletines = sum(Boletin),
    Vales = sum(Vale),
    .groups = "drop"
  )

write_csv(
  uso_mes,
  file.path(
    salida,
    "tablas",
    "uso_boletines_vales_por_mes.csv"
  )
)

uso_largo <- uso_mes |>
  pivot_longer(
    cols = c(Boletines, Vales),
    names_to = "Tipo",
    values_to = "Cantidad"
  )

p <- ggplot(
  uso_largo,
  aes(
    Mes,
    Cantidad,
    group = Tipo,
    linetype = Tipo
  )
) +
  geom_line() +
  geom_point() +
  labs(
    title = "Uso de boletines y vales por mes",
    x = "Mes",
    y = "Cantidad"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(
      angle = 45,
      hjust = 1
    )
  )

ggsave(
  file.path(
    salida,
    "graficos",
    "09_boletines_vales_por_mes.png"
  ),
  p,
  width = 10,
  height = 5,
  dpi = 160
)

# Correlaciones
pearson <- cor.test(
  df$Edad,
  df$Venta_total,
  method = "pearson"
)

tabla_genero_pago <- table(
  df$Genero_desc,
  df$MetodoPago_desc
)

tabla_boletin_vale <- table(
  df$Boletin_desc,
  df$Vale_desc
)

cramers_v <- function(tabla) {
  chi <- suppressWarnings(
    chisq.test(
      tabla,
      correct = FALSE
    )$statistic
  )

  n <- sum(tabla)
  dimensiones <- dim(tabla)
  denominador <- min(
    dimensiones[1] - 1,
    dimensiones[2] - 1
  )

  if (denominador <= 0) {
    return(NA_real_)
  }

  as.numeric(
    sqrt((chi / n) / denominador)
  )
}

v_genero_pago <- cramers_v(tabla_genero_pago)
v_boletin_vale <- cramers_v(tabla_boletin_vale)

mes_mayor <- ventas_mes[
  which.max(ventas_mes$Ventas),
]

mes_menor <- ventas_mes[
  which.min(ventas_mes$Ventas),
]

frecuencia_nav <- sort(
  table(df$Navegador_desc),
  decreasing = TRUE
)

navegador_mas <- names(frecuencia_nav)[1]
navegador_menos <- names(frecuencia_nav)[length(frecuencia_nav)]

total_efectivo <- sum(
  df$MontoCompra[df$MetodoPago == 0]
)

resumen <- c(
  "RESUMEN DEL ANÁLISIS",
  "====================",
  sprintf(
    "Mes con mayores ventas: %s | %.2f",
    mes_mayor$Mes,
    mes_mayor$Ventas
  ),
  sprintf(
    "Mes con menores ventas: %s | %.2f",
    mes_menor$Mes,
    mes_menor$Ventas
  ),
  paste(
    "Navegador/canal más utilizado:",
    navegador_mas
  ),
  paste(
    "Navegador/canal menos utilizado:",
    navegador_menos
  ),
  sprintf(
    "Total pagado en efectivo: %.2f",
    total_efectivo
  ),
  sprintf(
    "Pearson Edad vs Venta_total: r=%.4f, p=%.6f",
    unname(pearson$estimate),
    pearson$p.value
  ),
  sprintf(
    "Cramér V Genero vs MetodoPago: %.4f",
    v_genero_pago
  ),
  sprintf(
    "Cramér V Boletin vs Vale: %.4f",
    v_boletin_vale
  )
)

writeLines(
  resumen,
  file.path(salida, "resumen.txt"),
  useBytes = TRUE
)

message(
  "Análisis terminado: ",
  normalizePath(salida)
)

File: Practica 1/python/02_analisis_datos.py
from __future__ import annotations

import argparse
from pathlib import Path
import math

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import pearsonr, chi2_contingency


MAP_GENERO = {0: "Masculino", 1: "Femenino"}
MAP_PAGO = {
    0: "Efectivo",
    1: "Tarjeta de Crédito",
    2: "Tarjeta de Débito"
}
MAP_NAVEGADOR = {
    0: "Tienda Física",
    1: "Navegador 1",
    2: "Navegador 2",
    3: "Navegador 3",
    4: "Navegador 4"
}
MAP_SI_NO = {0: "No", 1: "Sí"}


def moda_serie(serie):
    m = serie.mode(dropna=True)
    return m.iloc[0] if not m.empty else np.nan


def cramers_v(tabla):
    chi2, _, _, _ = chi2_contingency(tabla)
    n = tabla.to_numpy().sum()

    if n == 0:
        return np.nan

    filas, columnas = tabla.shape
    denominador = min(filas - 1, columnas - 1)

    if denominador <= 0:
        return np.nan

    return math.sqrt((chi2 / n) / denominador)


def guardar(fig, ruta):
    fig.tight_layout()
    fig.savefig(ruta, dpi=160, bbox_inches="tight")
    plt.close(fig)


def main():
    parser = argparse.ArgumentParser(description="Análisis de ventas YA LIMPIAS")
    parser.add_argument(
        "csv_limpio",
        nargs="?",
        default="ventas_online_2025_limpio.csv"
    )
    parser.add_argument("--salida", default="salida_python")
    args = parser.parse_args()

    salida = Path(args.salida)
    graficos = salida / "graficos"
    tablas = salida / "tablas"

    graficos.mkdir(parents=True, exist_ok=True)
    tablas.mkdir(parents=True, exist_ok=True)

    # IMPORTANTE:
    # Aquí no se realiza limpieza. Se asume que el archivo ya fue
    # generado con 01_limpieza_datos.py.
    df = pd.read_csv(args.csv_limpio)
    df["FechaCompra"] = pd.to_datetime(df["FechaCompra"], format="%Y-%m-%d")

    df["Mes"] = df["FechaCompra"].dt.to_period("M").astype(str)
    df["Genero_desc"] = df["Genero"].map(MAP_GENERO)
    df["MetodoPago_desc"] = df["MetodoPago"].map(MAP_PAGO)
    df["Navegador_desc"] = df["Navegador"].map(MAP_NAVEGADOR)
    df["Boletin_desc"] = df["Boletin"].map(MAP_SI_NO)
    df["Vale_desc"] = df["Vale"].map(MAP_SI_NO)

    # 1. Estadísticas básicas
    columnas = ["Edad", "Venta_total", "N_Compras", "MontoCompra", "Tiempo"]

    estadisticas = []
    for columna in columnas:
        estadisticas.append({
            "Variable": columna,
            "Media": df[columna].mean(),
            "Mediana": df[columna].median(),
            "Moda": moda_serie(df[columna]),
            "Desviacion_estandar": df[columna].std(),
            "Minimo": df[columna].min(),
            "Maximo": df[columna].max()
        })

    pd.DataFrame(estadisticas).to_csv(
        tablas / "estadisticas_basicas.csv",
        index=False,
        encoding="utf-8-sig"
    )

    # 2. Ventas por mes
    ventas_mes = (
        df.groupby("Mes", as_index=False)["MontoCompra"]
        .sum()
        .rename(columns={"MontoCompra": "Ventas"})
    )
    ventas_mes.to_csv(tablas / "ventas_por_mes.csv", index=False, encoding="utf-8-sig")

    fig, ax = plt.subplots(figsize=(10, 5))
    ax.plot(ventas_mes["Mes"], ventas_mes["Ventas"], marker="o")
    ax.set_title("Ventas por mes")
    ax.set_xlabel("Mes")
    ax.set_ylabel("Monto de ventas")
    ax.tick_params(axis="x", rotation=45)
    guardar(fig, graficos / "01_ventas_por_mes.png")

    # 3. Método de pago
    pago = (
        df.groupby("MetodoPago_desc")["MontoCompra"]
        .sum()
        .sort_values(ascending=False)
    )
    pago.to_csv(tablas / "ventas_por_metodo_pago.csv", encoding="utf-8-sig")

    fig, ax = plt.subplots(figsize=(8, 5))
    pago.plot(kind="bar", ax=ax)
    ax.set_title("Ventas por método de pago")
    ax.set_xlabel("Método de pago")
    ax.set_ylabel("Monto de ventas")
    ax.tick_params(axis="x", rotation=20)
    guardar(fig, graficos / "02_ventas_por_metodo_pago.png")

    # 4. Navegador / canal
    navegador = (
        df.groupby("Navegador_desc")["MontoCompra"]
        .sum()
        .sort_values(ascending=False)
    )
    navegador.to_csv(tablas / "ventas_por_navegador.csv", encoding="utf-8-sig")

    fig, ax = plt.subplots(figsize=(9, 5))
    navegador.plot(kind="bar", ax=ax)
    ax.set_title("Ventas por navegador/canal")
    ax.set_xlabel("Navegador / canal")
    ax.set_ylabel("Monto de ventas")
    ax.tick_params(axis="x", rotation=25)
    guardar(fig, graficos / "03_ventas_por_navegador.png")

    # 5. Boletín
    boletin = df.groupby("Boletin_desc")["MontoCompra"].sum()
    boletin.to_csv(tablas / "ventas_por_boletin.csv", encoding="utf-8-sig")

    fig, ax = plt.subplots(figsize=(7, 5))
    boletin.plot(kind="bar", ax=ax)
    ax.set_title("Ventas según uso de boletín")
    ax.set_xlabel("Boletín")
    ax.set_ylabel("Monto de ventas")
    ax.tick_params(axis="x", rotation=0)
    guardar(fig, graficos / "04_ventas_por_boletin.png")

    # 6. Vale
    vale = df.groupby("Vale_desc")["MontoCompra"].sum()
    vale.to_csv(tablas / "ventas_por_vale.csv", encoding="utf-8-sig")

    fig, ax = plt.subplots(figsize=(7, 5))
    vale.plot(kind="bar", ax=ax)
    ax.set_title("Ventas según uso de vale")
    ax.set_xlabel("Vale")
    ax.set_ylabel("Monto de ventas")
    ax.tick_params(axis="x", rotation=0)
    guardar(fig, graficos / "05_ventas_por_vale.png")

    # 7. Correlación venta total vs edad
    fig, ax = plt.subplots(figsize=(8, 5))
    ax.scatter(df["Edad"], df["Venta_total"], alpha=0.6)
    ax.set_title("Venta total vs edad")
    ax.set_xlabel("Edad")
    ax.set_ylabel("Venta total")
    guardar(fig, graficos / "06_venta_total_vs_edad.png")

    # 8. Segmentación por edad
    intervalos = [0, 17, 25, 35, 45, 55, 65, np.inf]
    etiquetas = ["<=17", "18-25", "26-35", "36-45", "46-55", "56-65", "66+"]

    df["GrupoEdad"] = pd.cut(
        df["Edad"],
        bins=intervalos,
        labels=etiquetas,
        include_lowest=True
    )

    edad = (
        df.groupby("GrupoEdad", observed=False)["MontoCompra"]
        .agg(Compras="count", Ventas="sum", TicketPromedio="mean")
        .reset_index()
    )
    edad.to_csv(tablas / "segmentacion_por_edad.csv", index=False, encoding="utf-8-sig")

    fig, ax = plt.subplots(figsize=(9, 5))
    ax.bar(edad["GrupoEdad"].astype(str), edad["Ventas"])
    ax.set_title("Ventas por grupo de edad")
    ax.set_xlabel("Grupo de edad")
    ax.set_ylabel("Monto de ventas")
    guardar(fig, graficos / "07_ventas_por_grupo_edad.png")

    # 9. Comparación entre géneros
    genero = (
        df.groupby("Genero_desc")["MontoCompra"]
        .agg(Compras="count", Ventas="sum", TicketPromedio="mean")
        .reset_index()
    )
    genero.to_csv(tablas / "comparacion_genero.csv", index=False, encoding="utf-8-sig")

    fig, ax = plt.subplots(figsize=(7, 5))
    ax.bar(genero["Genero_desc"], genero["Ventas"])
    ax.set_title("Ventas por género")
    ax.set_xlabel("Género")
    ax.set_ylabel("Monto de ventas")
    guardar(fig, graficos / "08_ventas_por_genero.png")

    # 10. Uso de boletines y vales por mes
    uso_mes = (
        df.groupby("Mes", as_index=False)
        .agg(
            Boletines=("Boletin", "sum"),
            Vales=("Vale", "sum")
        )
    )
    uso_mes.to_csv(
        tablas / "uso_boletines_vales_por_mes.csv",
        index=False,
        encoding="utf-8-sig"
    )

    fig, ax = plt.subplots(figsize=(10, 5))
    ax.plot(uso_mes["Mes"], uso_mes["Boletines"], marker="o", label="Boletines")
    ax.plot(uso_mes["Mes"], uso_mes["Vales"], marker="o", label="Vales")
    ax.set_title("Uso de boletines y vales por mes")
    ax.set_xlabel("Mes")
    ax.set_ylabel("Cantidad")
    ax.legend()
    ax.tick_params(axis="x", rotation=45)
    guardar(fig, graficos / "09_boletines_vales_por_mes.png")

    # Correlaciones
    r_edad, p_edad = pearsonr(df["Edad"], df["Venta_total"])

    tabla_genero_pago = pd.crosstab(
        df["Genero_desc"],
        df["MetodoPago_desc"]
    )

    tabla_boletin_vale = pd.crosstab(
        df["Boletin_desc"],
        df["Vale_desc"]
    )

    tabla_genero_pago.to_csv(
        tablas / "contingencia_genero_metodo_pago.csv",
        encoding="utf-8-sig"
    )
    tabla_boletin_vale.to_csv(
        tablas / "contingencia_boletin_vale.csv",
        encoding="utf-8-sig"
    )

    v_genero_pago = cramers_v(tabla_genero_pago)
    v_boletin_vale = cramers_v(tabla_boletin_vale)

    mes_mayor = ventas_mes.loc[ventas_mes["Ventas"].idxmax()]
    mes_menor = ventas_mes.loc[ventas_mes["Ventas"].idxmin()]

    frecuencia_nav = df["Navegador_desc"].value_counts()
    navegador_mas = frecuencia_nav.idxmax()
    navegador_menos = frecuencia_nav.idxmin()

    total_efectivo = df.loc[
        df["MetodoPago"] == 0,
        "MontoCompra"
    ].sum()

    resumen = [
        "RESUMEN DEL ANÁLISIS",
        "====================",
        f"Mes con mayores ventas: {mes_mayor['Mes']} | {mes_mayor['Ventas']:.2f}",
        f"Mes con menores ventas: {mes_menor['Mes']} | {mes_menor['Ventas']:.2f}",
        f"Navegador/canal más utilizado: {navegador_mas}",
        f"Navegador/canal menos utilizado: {navegador_menos}",
        f"Total pagado en efectivo: {total_efectivo:.2f}",
        f"Pearson Edad vs Venta_total: r={r_edad:.4f}, p={p_edad:.6f}",
        f"Cramér V Genero vs MetodoPago: {v_genero_pago:.4f}",
        f"Cramér V Boletin vs Vale: {v_boletin_vale:.4f}",
    ]

    (salida / "resumen.txt").write_text(
        "\n".join(resumen),
        encoding="utf-8"
    )

    print(f"Análisis terminado: {salida.resolve()}")


if __name__ == "__main__":
    main()
